### PR TITLE
Support non-URC events for Quectel GSM Modules

### PR DESCRIPTION
**Pull request by Mtalat**
- This pull request updates the cellular library to support non-URC events sent by the modem, such as:   \<socket_index>,CONNECT OK. This is required for Quectel GSM Modules added in https://github.com/FreeRTOS/FreeRTOS-Cellular-Interface-Community-Supported-Ports/pull/3
- The unit tests are updated to include the changes

### CODE REVIEW

**Comment by awsblake** [comment]
Hi @Mtalat,

We suggest you move this port to the community support repo.
https://github.com/FreeRTOS/FreeRTOS-Cellular-Interface-Community-Supported-Ports

This repo will be existed in the module/ThridPart directory.

File: source/cellular_at_core.c [hunk 1]
```diff
@@ -105,6 +105,7 @@ CellularATError_t Cellular_ATIsPrefixPresent( const char * pString,
     CellularATError_t atStatus = CELLULAR_AT_SUCCESS;
     CellularATStringValidationResult_t stringValidationResult = CELLULAR_AT_STRING_UNKNOWN;
     char * ptrPrefixChar = NULL;
+    char * ptrPrefixChar_nonURC = NULL;
```
[the remark this answers is not in the document]
Comment (Mtalat) [reply source/cellular_at_core.c:108]: QIOPEN  for example, the modem responds with "OK" then the command output, as follows:

image

File: source/cellular_pkthandler.c [hunk 2]
```diff
@@ -155,6 +156,17 @@ static CellularPktStatus_t urcParseToken( CellularContext_t * pContext,
...
+    else if( *( ++pSavePtr ) == ',' )
+    {
+        pSavePtr++;
+
+        if( *pSavePtr == '\0' )
+        {
+            LogError( ( "_Cellular_AtParse : input string error, contains \",\" but no token %s", pInputLine ) );
+            pktStatus = CELLULAR_PKT_STATUS_BAD_REQUEST;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Mtalat) [reply source/cellular_pkthandler.c:169]: for socket opening and closing events, (0, CONNECT OK) or (0, CLOSE OK) for example, here we have the socket index as a variable prefix so it is not possible to use fixed URC tokens.

**Comment by chinglee-iot** [comment]
@Mtalat

I don't have the cellular modem. I use a similar example from the document. 

```
AT+QIOPEN=1,“TCP”,“220.180.239.212”,“8063” <= AT command
OK <= AT command SRC success token
1, CONNECT OK <= This line will be regarded as AT_UNDEFINED since it is not in CellularUrcTokenWoPrefixTable and it doesn't has prefix.
```

Can you help to try [this patch](https://github.com/chinglee-iot/FreeRTOS-Cellular-Interface/commit/28245b5e5bdb47062aecf8d0d5d776bf8eadf734)?

Also update the module porting.
https://github.com/1NCE-GmbH/FreeRTOS-Cellular-Interface-Community-Supported-Ports/blob/e882083635a88b6a6a33b4190b6e40515790d0fa/qgsm/cellular_qgsm.c#L109
```C
static bool prvUndefinedRespHandler( CellularContext_t * pContext, const char * pLine )
{
    bool result = true;

    LogInfo( ( "prvUndefinedRespHandler : input line %s", pLine ) );

    /* Handle the AT_UNKNOWN message here. If it can be handled, return true. Otherwise,
     * return false to pktio to indicate error.  "0, CONNECT OK" can be handled in the callback. */
    return result;
}

CellularError_t Cellular_ModuleInit( const CellularContext_t * pContext,
                                     void ** ppModuleContext )
{
    ...
    if( cellularStatus == CELLULAR_SUCCESS )
    {
        cellularStatus = _Cellular_RegisterUndefinedRespCallback( pContext, prvUndefinedRespHandler );
    }
    return cellularStatus;
}
```

If this patch can help your porting, I will try to merge the patch to this repository.

**Comment by chinglee-iot** [comment]
This PR is included in PR #71.
Please help to update your porting accordingly.
Many thanks to your contribution. I will close this PR first.
If there is any question, please re-open this PR and reply.

**Pull request closed without merging**